src/redismodule.h

```diff
@@ -807,6 +816,15 @@ typedef void (*RedisModuleScanCB)(RedisModuleCtx *ctx, RedisModuleString *keynam
 typedef void (*RedisModuleScanKeyCB)(RedisModuleKey *key, RedisModuleString *field, RedisModuleString *value, void *privdata);
 typedef void (*RedisModuleUserChangedFunc) (uint64_t client_id, void *privdata);
 typedef int (*RedisModuleDefragFunc)(RedisModuleDefragCtx *ctx);
+typedef RedisModuleString * (*RedisModuleConfigGetStringFunc)(const char *name, void *privdata);
+typedef long long (*RedisModuleConfigGetNumericFunc)(const char *name, void *privdata);
+typedef int (*RedisModuleConfigGetBoolFunc)(const char *name, void *privdata);
+typedef int (*RedisModuleConfigGetEnumFunc)(const char *name, void *privdata);
+typedef int (*RedisModuleConfigSetStringFunc)(const char *name, RedisModuleString *val, void *privdata, const char **err);
+typedef int (*RedisModuleConfigSetNumericFunc)(const char *name, long long val, void *privdata, const char **err);
+typedef int (*RedisModuleConfigSetBoolFunc)(const char *name, int val, void *privdata, const char **err);
+typedef int (*RedisModuleConfigSetEnumFunc)(const char *name, int val, void *privdata, const char **err);
+typedef int (*RedisModuleConfigApplyFunc)(RedisModuleCtx *ctx, void *privdata, const char **err);
```

Comment:
reviving this topic in a new thread, since the previous one is lost far above.
maybe we want a `RedisModuleString *err` so that modules can format the string and not have to rely on a static one.

Reply:
The contract can be that on our side is that we don't mutate or retain a pointer to it, just copy the value out. The caller is then free to pass back a reference it retains so that it can be freed or modified later. The reason configs use char *err instead of sds is because the vast majority of cases don't need special allocation or handling. I think the proposed approach would just introduce a lot more creating of strings.

Reply:
ok, you suggest that it'll remain a `char*`, and then the module can either use a static string, or some global pointer to a heap allocated one, which it releases next time / later.

@gkorland would that work for the non-C modules too?

Reply:
I actually like the original idea of passing a newly allocated RedisModuleString that Redis frees. It's simple, flexible and I don't see why it should be incompatible with non-C modules.

Reply:
allocating RedisModuleString is surely compatible with everything.
the question was if using a plain `char*` (without passing ownership) is acceptable.
at least for C modules, it's much more convenient since usually the errors are static string.
and as madolson suggested, modules can even use a global buffer, and format a message into it (as long as redis grantees to copy it right away)